### PR TITLE
Persistent selections for cartesian subplots

### PR DESCRIPTION
**Pull request by archmoj**
[Demo](https://codepen.io/MojtabaSamimi/pen/MWQMpya)

Please find new `layout` attributes in https://github.com/plotly/plotly.js/commit/b952125d9f5cf326d94a7831df4362e3214a31a9 which are quite similar to `shapes`.

More description: coming soon

Some TODOs:
 - [x] revise `x0`, `x1`, `y0`, `y1` defaults
 - [x] handle `splom`
 - [x] improve activating selections
 - [x] improve subtract cases & multi-polygon cases
 - [x] additional work on tests
 - [x] consider Chris' comments on the slack channel
 
 @plotly/plotly_js 
cc:  @chriddyp

### CODE REVIEW

File: src/components/selections/attributes.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+var dash = require('../drawing/attributes').dash;
+var extendFlat = require('../../lib/extend').extendFlat;
+var overrideAll = require('../../plot_api/edit_types').overrideAll;
+var templatedArray = require('../../plot_api/plot_template').templatedArray;
+var axisPlaceableObjs = require('../../constants/axis_placeable_objects');
+
+module.exports = overrideAll(templatedArray('selection', {
+    type: {
+        valType: 'enumerated',
+        values: ['rect', 'path'],
```
Comment (alexcjohnson) [inline src/components/selections/attributes.js:14]: `dflt: 'rect'`?

**Comment by alexcjohnson** [comment]
@archmoj this is looking great! A couple of bits of behavior I notice though:
- When I make two selections in a row, then click back onto the first selection and edit it, the second selection disappears.
- The first selection selects points immediately while the mouse move as it should. But the second shape doesn't select points even on mouse up, only when I click off it.
- Similarly when I modify an existing selection shape, the selected points don't change until mouse up - they should change while I'm dragging, with an associated `plotly_selecting` event.
- Currently if I make one selection, then shift-drag, it modifies that selection. If the new rectangle overlaps the old one it converts the combined shape into a path. I feel like the general flow should be closer to what it was before: a regular drag clears the previous selection and starts making a new one, shift-drag adds a new shape (but works like plain drag does today, ie the two shapes stay independent rather than combining into one)
- Do we really need the gray fill when a selection is selected? Aren't the handles enough? I feel like there will be a lot of situations where that obscures the data too much. Also the ghost "original shape" when you're modifying a selection - neat idea, could include it for editable shapes but I think it's too much for selections.
- This one I'd say we can do later, but I agree with @chriddyp's thought that we shouldn't have a state where no selection is active. When you create a selection, it should be active from mouseup. When you add another one, it should become the active one. If you then then click on the first one, that one should go back to being active.

**Comment by archmoj** [comment]
> * Do we really need the gray fill when a selection is selected? Aren't the handles enough? I feel like there will be a lot of situations where that obscures the data too much. Also the ghost "original shape" when you're modifying a selection - neat idea, could include it for editable shapes but I think it's too much for selections.

Yes. I think that's a pretty useful default namely to illustrate which regions of the selection are adding or removing.
One could simply tweak `activeselection.fillcolor` or `activeselection.opacity` in cases that it may not be desirable. E.g.
```js
Plotly.relayout(gd, 'activeselection.fillcolor', 'rgba(0,0,0,0)');
```

**Comment by archmoj** [comment]
> * The first selection selects points immediately while the mouse move as it should. But the second shape doesn't select points even on mouse up, only when I click off it.

Addressed mainly in https://github.com/plotly/plotly.js/pull/6243/commits/b3025e6927f245f69677fb5ff8208af3d55efe99.

**Comment by archmoj** [comment]
> * Do we really need the gray fill when a selection is selected? Aren't the handles enough? I feel like there will be a lot of situations where that obscures the data too much. Also the ghost "original shape" when you're modifying a selection - neat idea, could include it for editable shapes but I think it's too much for selections.

Addressed by https://github.com/plotly/plotly.js/pull/6243/commits/d28cb4f4e73ce1ffcbdde3393983972b7b6ca1eb.

**Comment by archmoj** [comment]
> * This one I'd say we can do later, but I agree with @chriddyp's thought that we shouldn't have a state where no selection is active. When you create a selection, it should be active from mouseup. When you add another one, it should become the active one. If you then then click on the first one, that one should go back to being active.

Activating a selection is simplified by https://github.com/plotly/plotly.js/pull/6243/commits/4138cf040cc5a0e661781b64aaeafc53b3961f3f which I believe help address this.

**Comment by archmoj** [comment]
This PR is ready for review.

**Comment by archmoj** [comment]
@alexcjohnson This PR is ready for your review. Thanks :pray:

File: src/components/selections/select.js [hunk 2]
```diff
@@ -377,6 +377,8 @@ function prepSelect(evt, startX, startY, dragOptions, mode) {
         throttle.done(throttleID).then(function() {
             throttle.clear(throttleID);
 
+            // Only points selected by the new selection are presented in eventData here
+            // Should we provide all the selected points instead?
```
Comment (alexcjohnson) [inline src/components/selections/select.js:381]: Yes, I think so 😄 (this is basically #4095, right?)
As it stands when you edit one of the selections later, the points from all selections are included; we should do that when you create a new selection as well.

**Comment by alexcjohnson** [comment]
Alright, getting close :slightly_smiling_face: Aside from my [latest comment](https://github.com/plotly/plotly.js/pull/6243#discussion_r918876485) (yes, always include all points in the event data) I only see two things:

(1) Currently when you’re in a selection mode and you click somewhere on the plot, whether you have a selection or not you get a selected event with no data. Can we just drop that event? ie if between mousedown and mouseup you didn't move the mouse enough to get a selecting event, don't emit a selected event.

(2) When you double click to deselect a region we don’t emit a `selected` event; we should, with just the remaining points. In fact when you double click to deselect everything, the first click gives an event (point 1 above, it shouldn't), but the second click which actually clears the selection doesn't give an event. It should, with no event data.

File: src/components/selections/select.js (hunk 2 above)
Comment (archmoj) [reply src/components/selections/select.js:381]: No, this `plotly_selected` one returns points similar to the `plotly_selecting` but it fires after a new selection not during the process of selecting. The shift/alt missing part is actually addressed in https://github.com/plotly/plotly.js/pull/6243/commits/b9d2db8788fb087d31a4a94b29332bb49d8a1065.

Comment (archmoj) [reply src/components/selections/select.js:404]: Addressed in https://github.com/plotly/plotly.js/pull/6243/commits/8776d49fcd10a0524304a95efe1a5626a6978d11.

**Comment by archmoj** [comment]
> (1) Currently when you’re in a selection mode and you click somewhere on the plot, whether you have a selection or not you get a selected event with no data. Can we just drop that event? ie if between mousedown and mouseup you didn't move the mouse enough to get a selecting event, don't emit a selected event.

It looks to be related to this part: 
https://github.com/plotly/plotly.js/blob/78f3700abd8a79578e545f059754f2fde381ef5a/src/components/selections/select.js#L385-L389
@alexcjohnson While I am out please feel free if you wanted to push commits here :)

File: src/components/selections/select.js [hunk 3]
```diff
@@ -1,27 +1,40 @@
 'use strict';
 
 var polybool = require('polybooljs');
+var pointInPolygon = require('point-in-polygon/nested');
```
Comment (alexcjohnson) [inline src/components/selections/select.js:4]: FYI we already have an "is this point inside this polygon" in [`polygon.tester`](https://github.com/plotly/plotly.js/blob/ee8523138145451ae6d5800cbc400e02f0c1d87f/src/lib/polygon.js#L159)

**Review by alexcjohnson — APPROVED** [review]
Excellent work! Let's do it 😄 💃

File: src/components/selections/select.js (hunk 3 above)
Comment (archmoj) [reply src/components/selections/select.js:4]: Good call. I tried to reuse it here: https://github.com/plotly/plotly.js/commit/f28576e44275dc6a4b2aea683a1a6b456b893dc7.
But it seems it require extra work.

Comment (alexcjohnson) [reply src/components/selections/select.js:4]: yeah it's old and undocumented... maybe I'll give it a shot myself at some point. Or maybe we drop that code entirely and just use `point-in-polygon` 😏

**Comment by etpinard** [comment]
Great feature - cheers :beers:

**Pull request merged**